src/rnnlm/sample_a_word.cc

```diff
@@ -0,0 +1,432 @@
+          if (order > 0) {
+            HistType::iterator last = h.end() - 1;
+            HistType::iterator first = h.begin() + 1;
+            HistType h1(h.begin(), last);
+            HistType h2(first, h.end());
+            prob = (it->second) * (pow(10, probs_[order][h][word].first) - 
+                    pow(10, GetBackoffWeight(order, h.back(), h1))
+                    * pow(10, GetProb(order, word, h2)));
+            (*pdf_w)[word] += prob;
+          } 
+          else {
+            prob = (it->second) * pow(10, probs_[order][h][word].first);
+            (*pdf_w)[word] += prob;
+          }
+        }
+      }
+    } // end reading history
+  } // end reading words
+}
+
+// sample a word that follows a pdf
+int32 NgramModel::SampleWord(const std::vector<float>& pdf) {
+  // generate a cdf from the given pdf 
+  std::vector<std::pair<float, float> > cdf;
+  float upper = 0;
+  float lower = 0;
+  std::pair<float, float> probs;
+  for (int32 i = 0; i < pdf.size(); i++) {
+    upper += pdf[i];
+    lower = upper - pdf[i];
+    probs = std::make_pair(lower, upper);
+    cdf.push_back(probs);
+  }
+  float u = 1.0 * rand()/RAND_MAX;
+  for (int32 i = 0; i < cdf.size(); i++) {
+    if (cdf[i].first <= u < cdf[i].second) {
+      return i;
+    }
+  }
+}
+
+// Sampling a word
+void NgramModel::TestSampling(int32 iters) {
+  ComputeHistoriesWeights();
+  std::vector<float> pdf;
+  ComputeWeightedPdf(&pdf);
+
+  // Compute diff
+  std::vector<float> pdf_est;
+  pdf_est.resize(vocab_size_);
+  int32 word;
+  int32 count_nons = 0;
+  for (int32 i = 0; i < iters; i++) {
+    word = SampleWord(pdf);
+    if (word > vocab_size_ || word < 0) {
+      std::cout << "the next word is " << word << std::endl;
+      count_nons += 1;
+      continue;
+    } else {
+      pdf_est[word] += 1.0;
+    }
+  }
+  // normalization
+  float ed = 0;
+  for (int32 i = 0; i < vocab_size_; i++) {
+    pdf_est[word] /= iters;
+    ed += pow(pdf_est[word] - pdf[word], 2); 
+  }
+  ed = pow(ed, 0.5);
+  std::cout << "Run " << iters << " times, e distance (expect < 0.05) is " << ed << std::endl;
+  std::cout << "Number of words OOV : " << count_nons << std::endl;
+}
+
+// Test the read-in language model
+void NgramModel::TestReadingModel() {
+  std::cout << "Testing model reading part..."<< std::endl;
+  std::cout << "Vocab size is: " << vocab_size_ << std::endl;
+  std::cout << "Ngram_order is: " << ngram_order_ << std::endl;
+  assert(probs_.size() == counts_.size());
+  for (int32 i = 0; i < ngram_order_; i++) {
+    int32 size_ngrams = 0;
+    std::cout << "Test: for order " << (i + 1) << std::endl;
+    std::cout << "Expected number of " << (i + 1) << "-grams: " << counts_[i] << std::endl;
+    for (auto it1 = probs_[i].begin(); it1 != probs_[i].end(); ++it1) {
+      HistType h(it1->first);
+      for (auto it2 = (probs_[i])[h].begin(); it2 != (probs_[i])[h].end(); ++it2) {
+        size_ngrams++; // number of words given
```

Comment:
use const HistType& if you don't change the key

m[iter->first] is always bad ==> use iter->second